### DISCUSSION TITLE
Add ability to customize a history server for different channels

### DISCUSSION THREAD

**Discussion by DmitryScaletta** [Chatterino/chatterino2#2586] [Ideas] [answered]
**What should be added?**

Ability to customize a history server for different channels.

The custom history server should return json like this 
```json
{
  "messages": []
}
```

**Why should it be added?**

Current [history server](https://recent-messages.robotty.de/) deletes all messages after 24 hours if there were no requests.

I want to have an ability to add my own history server for some channels.

Probably this feature won't be very popular and it probably don't need GUI in the settings.
I think it would be enough to add an option in `%appdata%/Chatterino2/Settings/settings.json` file. Something like this

```js
{
  /* ... */
  "historyServers": {
    "channel_name_1": "https://my-custom-history-server/recent-messages/%1",
    "channel_name_2": [
      "https://my-custom-history-server-1/recent-messages/%1",
      "https://my-custom-history-server-2/recent-messages/%1"
    ]
  }
}
```

If custom history server is not specified for the channel (or returns an error or 404) it uses the default one.

Related issues https://github.com/Chatterino/chatterino2/issues/1311 https://github.com/Chatterino/chatterino2/issues/982#issuecomment-478269419

**Comment by leon-richardt** [comment]
The provider for message history can be configured via the [`CHATTERINO2_RECENT_MESSAGES_URL`](https://wiki.chatterino.com/Environment%20Variables/#chatterino2_recent_messages_url) environment variable. Note that this is a global setting, not per channel.

Not sure about your exact use case but if you run your own message history server, you could provide your own logs for the channels you want to, and forward all other requests to https://recent-messages.robotty.de/ (given that @RAnders00 would be okay with that!).

**Comment by RAnders00** [comment]
Yes, you could probably solve this as well with a nginx config and by forwarding requests to your own instance or my public instance depending on channel name. That‘s a fine solution as far as I‘m concerned.

**Comment by DmitryScaletta** [comment]
So I added redirects to https://recent-messages.robotty.de/ in my custom history server on node.js (NestJS).

https://github.com/honeykingdom/honeykingdom-server/blob/0fd43f33f8d4429764fdf704c2e180942537730b/src/recent-messages/recent-messages.controller.ts#L26-L32

Added `CHATTERINO2_RECENT_MESSAGES_URL` variable.

But in chatterino history loads only for channels without redirects.
It didn't work for all other channels which my server redirects to https://recent-messages.robotty.de/

<details>
<summary>Redirect example</summary>
https://honeykingdom.herokuapp.com/api/v1/recent-messages/forsen
</details>

**Comment by leon-richardt** [comment]
Chatterino might not follow redirects when trying to fetch recent messages; I'm not sure about that at the moment. If it does not, that could be considered as a feature.

For the meanwhile, maybe fetch the logs from https://recent-messages.robotty.de/ and send them as a response to the request yourself. I'm aware that this is not an ideal solution but probably would be the fastest to get you up and running for now.

**Comment by RAnders00** [comment]
Are you using nginx as a reverse-proxy on your server? If yes, try this config:

```
server {
    listen 443 ssl http2;
    listen [::]:443 ssl http2;

    ssl_certificate /etc/letsencrypt/live/example.com/fullchain.pem;
    ssl_certificate_key /etc/letsencrypt/live/example.com/privkey.pem;
    server_name recent-messages.example.com;

    location ~ ^/api/v2/recent-messages/(?:channel_one|channel_two|channel_three|channel_four)$ {
        proxy_pass http://localhost:1234;
        include proxy_params;
    }

    location /api/v2/recent-messages {
        proxy_pass https://recent-messages.robotty.de;
        include proxy_params;
    }
}
```

Might need some tinkering to get working, I didn't test the config.

**Comment by RAnders00** [comment]
Also, although this is a bit off-topic, I saw you implemented the recent-messages service yourself, any reason why you didnt opt to run my implementation instead? (https://github.com/robotty/recent-messages2)

**Comment by DmitryScaletta** [comment]
> any reason why you didnt opt to run my implementation instead?

I wanted to write it on node.js.

**Comment by pajlada** [comment, marked as answer]
The added complexity does not seem worth it from Chatterino's perspective when this is a pretty easy solve from the server side (e.g. with Randers nginx config). Happy to continue this as a discussion instead as it still relates to Chatterino in a meaningful way
